From 7ae1b9336f8645daa0835d3a86d0520434acd913 Mon Sep 17 00:00:00 2001
From: Vanessa Maegima <vanessa.maegima@foundries.io>
Date: Mon, 12 Apr 2021 09:23:47 -0300
Subject: [PATCH 126/322] [FIO internal] imx8mm_evk: Dynamically set fdt_file
 based on board rev

Old board rev supports PMIC BD71837 while the latest rev supports
PCA9450, so probe this device to dynamically check the board revision
in order to set the correct fdt_file for each version.

To keep backward compatibility, the EVKB dtb was renamed to
imx8mm-evkb.dtb.

Signed-off-by: Vanessa Maegima <vanessa.maegima@foundries.io>
---
 board/freescale/imx8mm_evk/imx8mm_evk.c | 26 +++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/board/freescale/imx8mm_evk/imx8mm_evk.c b/board/freescale/imx8mm_evk/imx8mm_evk.c
index 93da5d81151..97cc671f033 100644
--- a/board/freescale/imx8mm_evk/imx8mm_evk.c
+++ b/board/freescale/imx8mm_evk/imx8mm_evk.c
@@ -298,6 +298,30 @@ int board_ehci_usb_phy_mode(struct udevice *dev)
 #define DISPMIX				9
 #define MIPI				10
 
+#define I2C_PMIC	0
+
+static void set_fdt_file(void)
+{
+	struct udevice *bus;
+	struct udevice *i2c_dev = NULL;
+	int ret;
+	uint8_t is_bd71837 = 0;
+
+	ret = uclass_get_device_by_seq(UCLASS_I2C, I2C_PMIC, &bus);
+	if (!ret)
+		ret = dm_i2c_probe(bus, 0x4b, 0, &i2c_dev);
+	if (!ret)
+		ret = dm_i2c_read(i2c_dev, 0x0, &is_bd71837, 1);
+
+	/* BD71837_REV, High Nibble is major version, fix 1010 */
+	is_bd71837 = !ret && ((is_bd71837 & 0xf0) == 0xa0);
+
+	if (!is_bd71837)
+		env_set("fdt_file", "imx8mm-evkb.dtb");
+	else
+		env_set("fdt_file", "imx8mm-evk-qca-wifi.dtb");
+}
+
 int board_init(void)
 {
 	struct arm_smccc_res res;
@@ -323,6 +347,8 @@ int board_late_init(void)
 	board_late_mmc_env_init();
 #endif
 
+	set_fdt_file();
+
 #ifdef CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG
 	env_set("board_name", "EVK");
 	env_set("board_rev", "iMX8MM");
-- 
2.34.1

